### PR TITLE
[dynamo] Desugar accumulate_grad, fix .grad handling

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120590
* #120520

Fixes https://github.com/pytorch/pytorch/issues/118435
Fixes https://github.com/pytorch/pytorch/issues/119906

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

File: torch/_dynamo/polyfill.py
Comment (ezyang) [inline torch/_dynamo/polyfill.py:47]: This is a simplification, are you planning to do the more complicated version?

Comment (anijain2305) [reply torch/_dynamo/polyfill.py:47]: I am but not in this PR. This is to unblock compiled ddp work. 

I need some more understanding and write new test cases with compiled_autograd to do a full fidelity solution. Maybe at the end of this week or next week.

Comment (ezyang) [reply torch/_dynamo/polyfill.py:47]: I guess the main trouble is that you are potentially changing the behavior; possibly evenly silently incorrect if x.grad is broadcasted and you need to un-broadcast it before adding the new grad in.

**Review by ezyang — APPROVED** [review]
Please check if the simplified accumulate grad polyfill implementation will cause problems, but I am loathe to block this PR because it does seem like the old thing was really busted and now it is much better!

File: torch/_dynamo/polyfill.py
Comment (jansel) [reply torch/_dynamo/polyfill.py:47]: I believe the grad_sum_to_size happens in compiled autograd not accumulate_grad_.  Though there is a lot of logic regarding layout this is missing.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**
